### PR TITLE
HHH-17096 treat() operator for @ManyToMany join generates SQL without discriminator condition

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-17096

### CODE REVIEW

**Comment by gavinking** [comment]
Thanks man.

File: hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java [hunk 1]
```diff
@@ -3100,8 +3100,12 @@ private void registerEntityNameUsage(
 				(s, existingUse) -> finalEntityNameUse.stronger( existingUse )
 		);
 
-		// Resolve the table reference for all types which we register an entity name use for
-		if ( actualTableGroup.isInitialized() ) {
+		// Resolve the table reference for all types which we register an entity name use for.
+		// Also, force table group initialization for treats when needed to ensure correct cardinality
+		final EntityNameUse.UseKind useKind = finalEntityNameUse.getKind();
+		if ( actualTableGroup.isInitialized() || ( useKind == EntityNameUse.UseKind.TREAT
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java:3106]: I guess `canUseInnerJoin()` might be a good indicator to understand if the treat is necessary to retain cardinality. Did you try that before?

```suggestion
		if ( actualTableGroup.isInitialized() || ( useKind == EntityNameUse.UseKind.TREAT && actualTableGroup.canUseInnerJoins()
```

Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java]: Yes checking `canUseInnerJoin()` makes sense, it should help avoid unnecessary initializations and type casts, I've added that to the condition - thanks.

By the way, the `instanceof EntityMappingType` check might be unnecessary, I added it to be safe.

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java]: Indeed, the `instanceof` check is unnecessary as this is only ever called for table groups for entity types.

Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java]: Gone :+1:

**Review by beikov — APPROVED** [review]

**Pull request merged**